### PR TITLE
Fix: add type check for `ArrayPhysicalCast` before downcasting 

### PR DESCRIPTION
**Pull request by HarshitaKalani**
Fixes: https://github.com/lfortran/lfortran/issues/4239
Fixes: https://github.com/lfortran/lfortran/issues/4062
Towards: #4017

### CODE REVIEW

File: integration_tests/intrinsics_250.f90 [hunk 1]
```diff
@@ -0,0 +1,17 @@
+program intrinsics_50
```
Comment (Pranavchiku) [inline integration_tests/intrinsics_250.f90:1]:
```suggestion
program intrinsics_250
```

**Review by Pranavchiku — APPROVED** [review]
changes are good, I'll request you to apply comments and add a new line at end of test files.

**Comment by HarshitaKalani** [comment]
>add a new line at end of test files.

Do you mean remove the new line? Since I've included a new line in the tests already.

File: integration_tests/intrinsics_249.f90 [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    contains
+    function var_2_iint8_dp(x, dim) result(res)
+    use iso_fortran_env, only: dp => real64
+    integer, intent(in) :: x(:,:)
+    integer, intent(in) :: dim
+    real(dp) :: res
+    res = sum(sum(abs(x), dim))
+    end function var_2_iint8_dp
+end program
+
```
Comment (Pranavchiku) [inline integration_tests/intrinsics_249.f90:20]:
```suggestion

```

File: integration_tests/intrinsics_250.f90 [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+function func(x, dim) result(res)
+integer, intent(in) :: x(:,:)
+integer, intent(in) :: dim
+real(8) :: res(2)
+print*, sum(abs(x), dim)
+res = sum(abs(x), dim)
+if (any(res /= [4, 4])) error stop
+end function func
+end program
+
```
Comment (Pranavchiku) [inline integration_tests/intrinsics_250.f90:17]:
```suggestion

```

**Comment by HarshitaKalani** [comment]
I'll just merge this in order to proceed with the completion of  #4063. I guess there's just one small issue left after this to complete the implementation of `real`.

**Review by Pranavchiku — APPROVED** [review]

**Pull request merged**